core/src/main/java/com/linecorp/armeria/client/routing/endpoint/healthcheck/HealthCheckedEndpointGroup.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.routing.endpoint.healthcheck;
+
+import static com.linecorp.armeria.common.util.Functions.voidFunction;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.IntStream;
+
+import com.google.common.collect.ImmutableList;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.client.ClientFactory;
+import com.linecorp.armeria.client.Endpoint;
+import com.linecorp.armeria.client.routing.EndpointGroup;
+
+import jp.skypencil.guava.stream.GuavaCollectors;
+
+/**
+ * An {@link EndpointGroup} decorator that only provides healthy {@link Endpoint}s.
+ */
+public abstract class HealthCheckedEndpointGroup implements EndpointGroup {
+    public static final long DEFAULT_HEALTHCHECK_RETRY_DELAY_MILLS = 3_000; // 3 seconds.
```

Comment:
I'm not sure if it's a good idea to make the default configuration value visible, mainly because it leaks the implementation detail and it potentially makes a user believe that the value won't change. How about making it a private value?